src/pyats/contrib/creators/topology.py

```diff
@@ -0,0 +1,656 @@
+                                    'os': os,
+                                    'finder': discover_name}
+        else:
+            if device_list[dest_host]['os'] is None:
+                device_list[dest_host]['os'] = os
+            device_list[dest_host]['ports'].add(dest_port)
+            device_list[dest_host]['ip'] = device_list[dest_host]['ip'].union(ip_address)
+
+    def add_to_connection_dict(self, connection_dict, 
+                               dest_host, dest_port, 
+                               ip_address,interface, dev):
+        '''
+            Adds the information about a connection to be added to the topology
+            recording what device interface combo is connected to the given
+            interface and ip address involved in the connection
+        '''
+        new_entry = {'dest_host': dest_host, 
+                    'dest_port': dest_port, 
+                    'ip_address': ip_address}
+        if interface not in connection_dict:
+            connection_dict[interface] = [new_entry]
+        else:
+            unique = True
+            for entry in connection_dict[interface]:
+
+                # check that the connection being added is unique
+                if (entry['dest_host'] == dest_host 
+                        and entry['dest_port'] == dest_port):
+                    unique = False
+                    break
+            if unique:
+                log.info('Connection device {} interface {} to' 
+                         ' device {} interface {} found'.format(dev, 
+                                                                interface, 
+                                                                dest_host, 
+                                                                dest_port))
+                connection_dict[interface].append(new_entry)
+                
+    def get_credentials_and_proxies(self, testbed):
+        '''
+        Takes a copy of the current credentials in the testbed 
+        for use in connecting to other devices
+        '''
+        credential_dict = {}
+        proxy_list = set()
+        for device in testbed['devices'].values():
+
+            # get all connections used in the testbed
+            if 'credentials' not in device:
+                continue
+            for cred in device['credentials']:
+                if cred not in credential_dict :
+                    credential_dict[cred] = dict(device['credentials'][cred])
+                elif device['credentials'][cred] not in credential_dict.values():
+                    credential_dict[cred + str(len(credential_dict))] = dict(device['credentials'][cred])
+
+            #get list of proxies used in connections
+            for connect in device['connections'].values():
+                if 'proxy' in connect:
+                    proxy_list.add(connect['proxy'])
+
+        return credential_dict, proxy_list
+
+    def get_os(self, system_string, platform_name):

```

Comment:
what's the `system_string` and `platform_name` here? for me, it looks like we can create device.api for this as well.